### PR TITLE
Processdupes

### PR DESCRIPTION
**Pull request by denismc88**
Changed order of duplicate processing to avoid duplication of work.

Simplified processduplicates to be config file driven, took code from removedups function to include logging.

### CODE REVIEW

File: code/processes/filealerter.q [hunk 1]
```diff
@@ -9,6 +9,7 @@ skipallonstart:@[value;`.fa.skipallonstart;0b]							// Whether to skip all acti
 moveonfail:@[value;`.fa.moveonfail;0b]								// If the processing of a file fails (by any action) then whether to move it or not regardless
 os:$[like[string .z.o;"w*"];`win;`lin]
 usemd5:@[value; `.fa.usemd5; 1b]								// Protected evaluation, returns value of usemd5 (from .fa namespace) or on fail, returns 1b
+processduplicate:@[value;`.fa.processduplicate;({x})]          // Run processduplicate from fa namespace on duplicate filenames passed in
```
Comment (jonnypress) [inline code/processes/filealerter.q]: This needs to be a {{x}} (a function which returns a function) rather than ({x}) (a function)
Otherwise the default definition will be set to be the result of {x} with the error string applied, so probably something like ".fa.processduplicate"

File: code/processes/filealerter.q [hunk 2]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
-
+    /-find all matches to the file search string
+    matches:find[.rmvr.removeenvvar[DICT[`path]];DICT[`match]];
+    toprocess:getunprocessed[matches;DICT[`newonly]];
+    files:exec filename from toprocess;
+    /-process duplicate files, either remove, move or leave depending on duplicateaction variable
+    processduplicate each except[matches;files];
+    /-If there are files to process
+    $[0<count files;
+    [{.lg.o[`alerter;"found file ", x]}'[files];
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: Can simplify this lambda to ``.lg.o'[`alerter;"found file ",/:files]``

File: code/processes/filealerter.q [hunk 3]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
+    /-perform the function on the file
+    pf:action/:[DICT[`function];files];];
+    .lg.o[`alerter;"no new files found"]];
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: This conditional is a bit hard to follow, may help to indent it.

File: code/processes/filealerter.q [hunk 4]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
+    matches:find[.rmvr.removeenvvar[DICT[`path]];DICT[`match]];
+    toprocess:getunprocessed[matches;DICT[`newonly]];
+    files:exec filename from toprocess;
+    /-process duplicate files, either remove, move or leave depending on duplicateaction variable
+    processduplicate each except[matches;files];
+    /-If there are files to process
+    $[0<count files;
+    [{.lg.o[`alerter;"found file ", x]}'[files];
+    /-perform the function on the file
+    pf:action/:[DICT[`function];files];];
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: Can index into `DICT` without the square brackets + I think there's an additional semi-colon that can be removed: `files]];`.

File: code/processes/filealerter.q [hunk 5]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
-        [{.lg.o[`alerter;"found file ", x]}'[files];
-        /-perform the function on the file
-        pf:action/:[DICT[`function];files];];
-        .lg.o[`alerter;"no new files found"]];
-        t:update function:(count toprocess)#DICT[`function],funcpassed:pf,moveto:(count toprocess)#enlist .rmvr.removeenvvar[DICT[`movetodirectory]] from toprocess; t}
-
-
+    /-find all matches to the file search string
+    matches:find[.rmvr.removeenvvar[DICT[`path]];DICT[`match]];
+    toprocess:getunprocessed[matches;DICT[`newonly]];
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: For neatness you can index into `DICT` without the squares, here and above.

File: code/processes/filealerter.q [hunk 6]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
-        $[0<count files;
-        [{.lg.o[`alerter;"found file ", x]}'[files];
-        /-perform the function on the file
-        pf:action/:[DICT[`function];files];];
-        .lg.o[`alerter;"no new files found"]];
-        t:update function:(count toprocess)#DICT[`function],funcpassed:pf,moveto:(count toprocess)#enlist .rmvr.removeenvvar[DICT[`movetodirectory]] from toprocess; t}
-
-
+    /-find all matches to the file search string
+    matches:find[.rmvr.removeenvvar[DICT[`path]];DICT[`match]];
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: Think ``find .rmvr.removeenvvar . DICT`path`match`` would work here.

File: code/processes/filealerter.q [hunk 7]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
+    toprocess:getunprocessed[matches;DICT[`newonly]];
+    files:exec filename from toprocess;
+    /-process duplicate files, either remove, move or leave depending on duplicateaction variable
+    processduplicate each except[matches;files];
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: Can use infix notation: `matches except files`

File: code/processes/filealerter.q [hunk 8]
```diff
@@ -93,19 +94,19 @@ chktable:{[files] table:([]filename:files;md5hash:$[usemd5;gethash'[files];(coun
...
+    /-If there are files to process
+    $[0<count files;
+    [{.lg.o[`alerter;"found file ", x]}'[files];
+    /-perform the function on the file
+    pf:action/:[DICT[`function];files];];
+    .lg.o[`alerter;"no new files found"]];
+    t:update function:(count toprocess)#DICT[`function],funcpassed:pf,moveto:(count toprocess)#enlist .rmvr.removeenvvar[DICT[`movetodirectory]] from toprocess; t}
```
Comment (ThomasSmyth) [inline code/processes/filealerter.q]: Use square brackets over parentheses: `count[...]`. `t` can also be returned directly without defining it `:update ... from toprocess;` + put closing brace `}` on new line.

File: config/settings/filealerter.q [hunk 9]
```diff
@@ -9,3 +9,4 @@ alreadyprocessed:first .proc.getconfigfile["filealerterprocessed"]	// The locati
 skipallonstart:0b						// Whether to skip all actions when the file alerter process starts up (so only "new" files after the processes starts will be processed)
 moveonfail:0b							// If the processing of a file fails (by any action) then whether to move it or not regardless
 usemd5:1b 							// User configuration for whether to find the md5 hash of new files. usemd5 takes 1b (on) or 0b (off)
+processduplicates:{.lg.o[`duplicates;"deleting duplicate file ",x];@[hdel;hsym`$x;{.lg.e[`duplicates;"failed to delete ",x," : ",y]}[x]]}
```
Comment (jonnypress) [inline config/settings/filealerter.q]: By default I think this definition should be empty, the same as it is at the top of the code/processes/filealerter.q
